src/buffer.rs

```diff
@@ -0,0 +1,301 @@
+                    self.high_buffer.clear();
+                    return SendResult::Disconnect;
+                }
+            }
+        }
+        while let Some(event) = self.normal_buffer.pop_front() {
+            match self.sender.poll_ready(cx) {
+                Poll::Ready(Ok(())) => {
+                    if let Err(e) = self.sender.try_send(event) {
+                        if e.is_full() {
+                            self.normal_buffer.push_front(e.into_inner());
+                            return SendResult::Pending;
+                        } else {
+                            return SendResult::Disconnect;
+                        }
+                    }
+                }
+                Poll::Pending => {
+                    self.normal_buffer.push_front(event);
+                    return SendResult::Pending;
+                }
+                Poll::Ready(Err(_)) => {
+                    self.normal_buffer.clear();
+                    return SendResult::Disconnect;
+                }
+            }
+        }
+        self.shrink_to_fit();
+        SendResult::Ok
+    }
+}
+
+pub struct Buffer<T> {
+    sender: Sender<T>,
+    buffer: VecDeque<T>,
+}
+
+impl<T> Buffer<T> {
+    pub fn new(sender: Sender<T>) -> Self {
+        Buffer {
+            sender,
+            buffer: VecDeque::default(),
+        }
+    }
+
+    pub fn push(&mut self, item: T) {
+        self.buffer.push_back(item)
+    }
+
+    pub fn len(&self) -> usize {
+        self.buffer.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.buffer.is_empty()
+    }
+
+    fn shrink_to_fit(&mut self) {
+        if self.buffer.capacity() > self.buffer.len() + BUF_SHRINK_THRESHOLD {
+            self.buffer.shrink_to_fit();
+        }
+    }
+
+    pub fn try_send(&mut self, cx: &mut Context) -> SendResult {
+        while let Some(event) = self.buffer.pop_front() {
+            match self.sender.poll_ready(cx) {
+                Poll::Ready(Ok(())) => {
+                    if let Err(e) = self.sender.try_send(event) {
+                        if e.is_full() {
+                            self.buffer.push_front(e.into_inner());
+                            return SendResult::Pending;
+                        } else {
+                            return SendResult::Disconnect;
+                        }
+                    }
+                }
+                Poll::Pending => {
+                    self.buffer.push_front(event);
+                    return SendResult::Pending;
+                }
+                Poll::Ready(Err(_)) => {
+                    self.buffer.clear();
+                    return SendResult::Disconnect;
+                }
+            }
+        }
+        self.shrink_to_fit();
+        SendResult::Ok
+    }
+
+    pub fn take(&mut self) -> (Sender<T>, VecDeque<T>) {
+        (self.sender.clone(), ::std::mem::take(&mut self.buffer))
+    }
+
+    pub fn sender(&self) -> Sender<T> {
```

Comment:
It's better to explicitly show that this function clones the sender.

```suggestion
    pub fn clone_sender(&self) -> Sender<T> {
```